src/main/java/appeng/integration/modules/waila/part/P2PStateWailaDataProvider.java

```diff
@@ -0,0 +1,158 @@
+
+import com.google.common.collect.Iterators;
+
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import mcp.mobius.waila.api.IWailaConfigHandler;
+import mcp.mobius.waila.api.IWailaDataAccessor;
+
+import appeng.api.parts.IPart;
+import appeng.core.localization.WailaText;
+import appeng.me.GridAccessException;
+import appeng.parts.p2p.PartP2PTunnel;
+
+
+/**
+ * Provides information about a P2P tunnel to WAILA.
+ */
+public final class P2PStateWailaDataProvider extends BasePartWailaDataProvider
+{
+
+	private static final int STATE_UNLINKED = 0;
+	private static final int STATE_OUTPUT = 1;
+	private static final int STATE_INPUT = 2;
+	public static final String TAG_P2P_STATE = "p2p_state";
+
+	/**
+	 * Adds state to the tooltip
+	 *
+	 * @param part part with state
+	 * @param currentToolTip to be added to tooltip
+	 * @param accessor wrapper for various information
+	 * @param config config settings
+	 *
+	 * @return modified tooltip
+	 */
+	@Override
+	public List<String> getWailaBody( final IPart part, final List<String> currentToolTip, final IWailaDataAccessor accessor, final IWailaConfigHandler config )
+	{
+		if( part instanceof PartP2PTunnel )
+		{
+			NBTTagCompound nbtData = accessor.getNBTData();
+			if( nbtData.hasKey( TAG_P2P_STATE ) )
+			{
+				int[] stateArr = nbtData.getIntArray( TAG_P2P_STATE );
+				if( stateArr.length == 2 )
+				{
+					int state = stateArr[0];
+					int outputs = stateArr[1];
+
+					switch( state )
+					{
+						case STATE_UNLINKED:
+							currentToolTip.add( WailaText.P2PUnlinked.getLocal() );
+							break;
+						case STATE_OUTPUT:
+							currentToolTip.add( WailaText.P2POutput.getLocal() );
+							break;
+						case STATE_INPUT:
+							currentToolTip.add( getOutputText( outputs ) );
+							break;
+					}
+				}
+			}
+		}
+
+		return currentToolTip;
+	}
+
+	@Override
+	public NBTTagCompound getNBTData( EntityPlayerMP player, IPart part, TileEntity te, NBTTagCompound tag, World world, BlockPos pos )
+	{
+		if( part instanceof PartP2PTunnel )
+		{
+			PartP2PTunnel tunnel = (PartP2PTunnel) part;
+
+			// The default state
+			int state = STATE_UNLINKED;
+			int outputCount = 0;
+
+			if( !tunnel.isOutput() )
+			{
+				outputCount = getOutputCount( tunnel );
+				if( outputCount > 0 )
+				{
+					// Only set it to INPUT if we know there are any outputs
+					state = STATE_INPUT;
+				}
+			}
+			else
+			{
+				PartP2PTunnel input = tunnel.getInput();
+				if( input != null )
+				{
+					state = STATE_OUTPUT;
+				}
+			}
+
+			tag.setIntArray( TAG_P2P_STATE, new int[] {
+					state,
+					outputCount
+			} );
+		}
+
+		return tag;
+	}
+
+	private static int getOutputCount( PartP2PTunnel tunnel )
+	{
+		try
+		{
+			return Iterators.size( tunnel.getOutputs().iterator() );
+		}
+		catch( GridAccessException e )
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)